File: android/app/src/main/kotlin/com/macci/kaalerto/data/Reducer.kt
package com.macci.kaalerto.data

import kotlin.math.exp
import kotlin.math.sqrt

/** Everything the map and detail sheet need to render one flooded spot. */
data class FeatureSummary(
    val featureRef: String,
    val lat: Double,
    val lon: Double,
    /** S0..S3, or "SX" for Rule C conflict. */
    val severity: String,
    val confidence: Double,
    val bucket: String,
    val isConflicted: Boolean,
    val lastEventMs: Long,
    val isStale: Boolean,
    val confirmCount: Int,
    val disputeCount: Int,
    /** Full history for this feature, most recent first — the detail sheet's list. */
    val events: List<Event>,
    /**
     * The severity an official has posted, whether or not it is currently in force.
     * Non-null and *not* equal to [severity] means the second-official gate is holding
     * it back — see [pendingSecondOfficial].
     */
    val officialSeverity: String? = null,
    /** Who posted it, and when. An official action is never anonymous (OfficialVerify.dc.html). */
    val officialAuthorName: String? = null,
    val officialAtMs: Long? = null,
    /**
     * An official has asked to lower a contradicted spot and is waiting for a second
     * official to agree. Day 10's gate — see [REQUIRED_OFFICIALS_TO_DEESCALATE].
     */
    val pendingSecondOfficial: Boolean = false,
    /**
     * How many residents' latest reports are *worse* than the official severity. Drives
     * OfficialReverse.dc.html's "3 residente ang salungat dito" and the detail sheet's
     * "N residents report worse conditions than the official status".
     */
    val contradictingCount: Int = 0,
)

private const val PROXIMATE_METERS = 500.0
private const val WEIGHT_FLOOR = 0.05
private const val CONFLICT_WEIGHT_THRESHOLD = 0.5
private const val DEESCALATION_COUNT = 2

/**
 * Day 10's second-official gate (OfficialReverse.dc.html: "Magkasalungat ang lugar na
 * ito, kaya hindi kayang ibaba ng iisang opisyal ang severity").
 *
 * It is the same safety asymmetry the crowd path already runs on — raising a severity
 * takes one credible voice, lowering one takes corroboration — extended to officials so
 * that a single person cannot quietly overrule residents who are standing in the water.
 * It applies **only** when lowering, and **only** when at least [DEESCALATION_COUNT]
 * residents are currently reporting worse. Raising a severity, and reversing another
 * official's clearance, stay single-official and immediate.
 */
private const val REQUIRED_OFFICIALS_TO_DEESCALATE = 2

private data class Weighted(val event: Event, val severity: String, val weight: Double, val proximate: Boolean)
private data class Resolution(
    val severity: String,
    val conflicted: Boolean,
    val bucket: String,
    val confidence: Double,
    /** The highest severity any weighted report claims, before Rules B/C resolve it. */
    val maxSeverity: String = severity,
)

/**
 * Pure fold over one feature's events into a display state — `state(feature, now) =
 * reduce(events_for(feature), now)` (docs/03-architecture.md §4.1). Deliberately a
 * simplified reading of that section: role-based weight instead of signed reputation
 * certificates (no crypto in this build, per ground rule 4), no independence/sybil
 * discounting, no sensor tier. BUILD_TASKS.md day 4's own literal spec is "weight =
 * role × proximity × exponential time decay" plus Rules B and C — this implements
 * exactly that, using docs/03-architecture.md §5.2-5.3's concrete formulas (proximity
 * bands, decay, Wilson score confidence) where they don't require that missing
 * infrastructure.
 */
object Reducer {

    fun summarize(featureRef: String, allEvents: List<Event>, now: Long): FeatureSummary? {
        val events = allEvents.filter { it.featureRef == featureRef }
        if (events.isEmpty()) return null

        // docs/03-architecture.md §4.3 step 4: one live position per author, so nobody
        // stacks repeated confirms to fake consensus.
        val latestPerAuthor = events.groupBy { it.authorId }.map { (_, authored) -> authored.maxBy { it.timestampMs } }

        // Anchor point for proximity: the original reports' centroid, not every
        // confirm/dispute — a confirm/dispute's own lat/lon is the CONFIRMING device's
        // location, not the flood's, so it must not shift where "the feature" is.
        val reports = events.filter { it.type == "flood_report" }
        val anchorLat = reports.map { it.lat }.average().takeUnless { it.isNaN() } ?: events.first().lat
        val anchorLon = reports.map { it.lon }.average().takeUnless { it.isNaN() } ?: events.first().lon

        val weighted = latestPerAuthor.mapNotNull { event ->
            val severity = event.severity ?: return@mapNotNull null
            val ageMs = (now - event.timestampMs).coerceAtLeast(0)
            val distance = haversineMeters(anchorLat, anchorLon, event.lat, event.lon)
            val proximity = proximityFactor(distance)
            val weight = roleWeight(event.authorRole) * proximity * recencyFactor(ageMs, severity)
            Weighted(event, severity, weight, proximity >= 0.7)
        }
        // The crowd is the crowd: an official's own ruling must not also be counted as
        // a report *within* the crowd it is overriding. Leaving it in made a lone
        // official clearance read as a two-sided disagreement and rendered SX — the
        // official arguing with the residents, at role weight 5.
        val crowdWeighted = weighted.filter { it.event.authorRole != "official" }
        val weightBySeverity = crowdWeighted.groupBy { it.severity }.mapValues { (_, ws) -> ws.sumOf { it.weight } }

        val crowd = resolveCrowd(crowdWeighted, weightBySeverity)

        // Rule D — an official event overrides the crowd, with day 10's one exception.
        val officialEvents = events.filter { it.authorRole == "official" && it.severity != null }
        val latestOfficial = officialEvents.maxByOrNull { it.timestampMs }
        val officialSeverity = latestOfficial?.severity

        // Residents currently saying it is *worse* than the official line. Officials are
        // excluded: this counts the people being overruled, not the people overruling.
        // "Still live" is the reducer's existing notion: weight above the floor, which
        // is time decay doing the work. Yesterday's reports have decayed away and do not
        // hold an all-clear hostage; people reporting worse right now do.
        val contradicting = officialSeverity?.let { official ->
            crowdWeighted.count {
                severityOrdinal(it.severity) > severityOrdinal(official) && it.weight > WEIGHT_FLOOR
            }
        } ?: 0

        // The gate only ever holds back a de-escalation of a contradicted spot. Raising a
        // severity, and reversing another official's clearance, are single-official.
        val isDeEscalation = officialSeverity != null &&
            severityOrdinal(officialSeverity) < severityOrdinal(crowd.maxSeverity)
        val gated = isDeEscalation && contradicting >= DEESCALATION_COUNT
        val backingOfficials = officialEvents
            .filter { it.severity == officialSeverity }
            .map { it.authorId }
            .distinct()
            .size
        val officialInForce = latestOfficial != null &&
            (!gated || backingOfficials >= REQUIRED_OFFICIALS_TO_DEESCALATE)

        val resolution = if (officialInForce) {
            resolveOfficial(latestOfficial!!, weightBySeverity)
        } else {
            crowd
        }

        return FeatureSummary(
            featureRef = featureRef,
            lat = anchorLat,
            lon = anchorLon,
            severity = resolution.severity,
            confidence = resolution.confidence,
            bucket = resolution.bucket,
            isConflicted = resolution.conflicted,
            lastEventMs = events.maxOf { it.timestampMs },
            isStale = now > events.maxOf { it.expiresAt },
            confirmCount = events.count { it.type == "confirm" },
            disputeCount = events.count { it.type == "dispute" },
            officialSeverity = officialSeverity,
            officialAuthorName = latestOfficial?.authorName,
            officialAtMs = latestOfficial?.timestampMs,
            pendingSecondOfficial = latestOfficial != null && gated && !officialInForce,
            contradictingCount = contradicting,
            events = events.sortedByDescending { it.timestampMs },
        )
    }

    fun summarizeAll(allEvents: List<Event>, now: Long): List<FeatureSummary> =
        allEvents.mapNotNull { it.featureRef }.distinct().mapNotNull { summarize(it, allEvents, now) }

    private fun resolveOfficial(officialEvent: Event, weightBySeverity: Map<String, Double>): Resolution {
        val severity = officialEvent.severity!!
        val total = weightBySeverity.values.sum()
        val agree = weightBySeverity[severity] ?: 0.0
        return Resolution(severity, conflicted = false, bucket = "official", confidence = if (total > 0) agree / total else 1.0)
    }

    private fun resolveCrowd(weighted: List<Weighted>, weightBySeverity: Map<String, Double>): Resolution {
        val severitiesWithWeight = weightBySeverity.filterValues { it > WEIGHT_FLOOR }.keys
        if (severitiesWithWeight.isEmpty()) return Resolution("S0", conflicted = false, bucket = "unverified", confidence = 0.0)

        // Rule A: the highest claimed tier is the ceiling — a single credible higher
        // report always wins over any number of older/lower ones.
        val maxSeverity = severitiesWithWeight.maxBy { severityOrdinal(it) }

        // Rule B: only a lower tier with >=2 proximate observers can pull the display
        // down from that high-water mark.
        val lowerEligible = severitiesWithWeight
            .filter { severityOrdinal(it) < severityOrdinal(maxSeverity) }
            .filter { sev -> weighted.count { it.severity == sev && it.proximate } >= DEESCALATION_COUNT }

        val dangerWeight = (weightBySeverity["S2"] ?: 0.0) + (weightBySeverity["S3"] ?: 0.0)
        val safeWeight = (weightBySeverity["S0"] ?: 0.0) + (weightBySeverity["S1"] ?: 0.0)
        // Rule C: genuine disagreement is shown, not resolved by count or averaging —
        // but a met de-escalation bar (Rule B) is a real resolution, so it pre-empts SX.
        val conflicted = lowerEligible.isEmpty() && dangerWeight >= CONFLICT_WEIGHT_THRESHOLD && safeWeight >= CONFLICT_WEIGHT_THRESHOLD

        val resolvedSeverity = when {
            conflicted -> "SX"
            lowerEligible.isNotEmpty() -> lowerEligible.maxBy { severityOrdinal(it) }
            else -> maxSeverity
        }

        if (conflicted) return Resolution(resolvedSeverity, conflicted = true, bucket = "unverified", confidence = 0.0, maxSeverity = maxSeverity)

        // docs/03-architecture.md §5.3: "Unverified: confidence < 0.35, OR a single
        // report" — the OR matters, since a lone report's Wilson bound can round to
        // just above 0.35 depending on weight; a single observer must never read as
        // more than Unverified regardless.
        if (weighted.size <= 1) {
            return Resolution(
                resolvedSeverity,
                conflicted = false,
                bucket = "unverified",
                confidence = weightBySeverity[resolvedSeverity]?.let { wilsonLowerBound(it, 0.0) } ?: 0.0,
                maxSeverity = maxSeverity,
            )
        }

        val agree = weightBySeverity[resolvedSeverity] ?: 0.0
        val disagree = weightBySeverity.filterKeys { it != resolvedSeverity }.values.sum()
        val confidence = wilsonLowerBound(agree, disagree)
        val proximateAgreeing = weighted.count { it.severity == resolvedSeverity && it.proximate }
        val bucket = when {
            confidence >= 0.65 && proximateAgreeing >= 2 -> "confirmed"
            confidence >= 0.35 -> "likely"
            else -> "unverified"
        }
        return Resolution(resolvedSeverity, conflicted = false, bucket = bucket, confidence = confidence, maxSeverity = maxSeverity)
    }
}

private fun roleWeight(role: String): Double = when (role) {
    "official" -> 5.0
    "responder" -> 2.5
    else -> 1.0
}

/** docs/03-architecture.md §5.2 proximity table, collapsed to the bands day 4 needs. */
private fun proximityFactor(distanceMeters: Double): Double = when {
    distanceMeters <= 100.0 -> 1.0
    distanceMeters <= PROXIMATE_METERS -> 0.7
    distanceMeters <= 2000.0 -> 0.4
    else -> 0.2
}

private fun recencyFactor(ageMs: Long, severity: String): Double {
    val ageMinutes = ageMs / 60_000.0
    return exp(-ageMinutes / decayTauMinutesFor(severity))
}

/** docs/03-architecture.md §5.3 — Wilson score lower bound on weighted counts, z≈1.44 (~85%). */
private fun wilsonLowerBound(agree: Double, disagree: Double): Double {
    val n = agree + disagree
    if (n <= 0.0) return 0.0
    val z = 1.44
    val z2 = z * z
    val pHat = agree / n
    val numerator = pHat + z2 / (2 * n) - z * sqrt(pHat * (1 - pHat) / n + z2 / (4 * n * n))
    val denominator = 1 + z2 / n
    return (numerator / denominator).coerceIn(0.0, 1.0)
}

File: dashboard/src/lib/reducer.ts
// A port of the phone's reducer (android/.../data/Reducer.kt, Withdraw.kt, Severity.kt, GeoUtil.kt),
// so the dashboard shows the same status for a spot as a phone holding the same events (NFR-4).
// Rules A–D, role x proximity x decay weighting, Wilson confidence and the buckets are copied
// line for line, with the same constants. Keep the two in step: `npm run check` runs the phone's
// reducer test cases (checks/reducer.check.mts) against this file.
//
// Deliberately imports only types, so the check script can run it under plain Node.
import type { Event } from './types';

export type Bucket = 'confirmed' | 'likely' | 'unverified' | 'official';

/** Everything the dashboard needs to render one flooded spot. Mirrors the phone's FeatureSummary. */
export interface FeatureSummary {
  featureRef: string;
  lat: number;
  lon: number;
  /** S0..S3, or "SX" for a Rule C conflict. */
  severity: string;
  confidence: number;
  bucket: Bucket;
  isConflicted: boolean;
  lastEventMs: number;
  isStale: boolean;
  confirmCount: number;
  disputeCount: number;
  /** Full history for this spot, most recent first, including withdrawals. */
  events: Event[];
  officialSeverity: string | null;
  officialAuthorName: string | null;
  officialAtMs: number | null;
  /** An official asked to lower a contradicted spot and is waiting for a second official. */
  pendingSecondOfficial: boolean;
  /** Residents whose latest report is worse than the official severity. */
  contradictingCount: number;
}

const PROXIMATE_METERS = 500;
const WEIGHT_FLOOR = 0.05;
const CONFLICT_WEIGHT_THRESHOLD = 0.5;
const DEESCALATION_COUNT = 2;
const REQUIRED_OFFICIALS_TO_DEESCALATE = 2;
const EARTH_RADIUS_M = 6_371_000;

const ORDER = ['S0', 'S1', 'S2', 'S3'];
/** S0..S3 low to high. "SX" is deliberately not on the ladder (ordinal 0), as on the phone. */
export const severityOrdinal = (s: string) => Math.max(0, ORDER.indexOf(s));

/** Severity-dependent TTL in minutes; also the decay time constant tau. */
const ttlMinutesFor = (s: string | null) => (s === 'S1' ? 120 : s === 'S2' ? 240 : s === 'S3' ? 360 : 60);

const TYPE_FLOOD_WITHDRAW = 'flood_withdraw';

export function haversineMeters(lat1: number, lon1: number, lat2: number, lon2: number): number {
  const rad = (d: number) => (d * Math.PI) / 180;
  const dLat = rad(lat2 - lat1);
  const dLon = rad(lon2 - lon1);
  const a = Math.sin(dLat / 2) ** 2 + Math.cos(rad(lat1)) * Math.cos(rad(lat2)) * Math.sin(dLon / 2) ** 2;
  return EARTH_RADIUS_M * 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
}

const roleWeight = (role: string) => (role === 'official' ? 5 : role === 'responder' ? 2.5 : 1);

const proximityFactor = (m: number) => (m <= 100 ? 1 : m <= PROXIMATE_METERS ? 0.7 : m <= 2000 ? 0.4 : 0.2);

const recencyFactor = (ageMs: number, severity: string) => Math.exp(-(ageMs / 60_000) / ttlMinutesFor(severity));

/** Wilson score lower bound on weighted counts, z about 1.44 (85%). */
function wilsonLowerBound(agree: number, disagree: number): number {
  const n = agree + disagree;
  if (n <= 0) return 0;
  const z = 1.44;
  const z2 = z * z;
  const pHat = agree / n;
  const numerator = pHat + z2 / (2 * n) - z * Math.sqrt((pHat * (1 - pHat)) / n + z2 / (4 * n * n));
  const denominator = 1 + z2 / n;
  return Math.min(1, Math.max(0, numerator / denominator));
}

/** Each author's events newer than their own latest withdrawal, and never the withdrawals themselves. */
export function liveEvents(featureEvents: Event[]): Event[] {
  const withdrawnAt = new Map<string, number>();
  for (const e of featureEvents) {
    if (e.type === TYPE_FLOOD_WITHDRAW) withdrawnAt.set(e.authorId, Math.max(withdrawnAt.get(e.authorId) ?? -Infinity, e.timestampMs));
  }
  return featureEvents.filter((e) => e.type !== TYPE_FLOOD_WITHDRAW && e.timestampMs > (withdrawnAt.get(e.authorId) ?? -Infinity));
}

interface Weighted {
  event: Event;
  severity: string;
  weight: number;
  proximate: boolean;
}

interface Resolution {
  severity: string;
  conflicted: boolean;
  bucket: Bucket;
  confidence: number;
  /** The highest severity any weighted report claims, before Rules B and C resolve it. */
  maxSeverity: string;
}

const maxBy = <T>(xs: T[], key: (x: T) => number): T | undefined =>
  xs.reduce<T | undefined>((best, x) => (best === undefined || key(x) > key(best) ? x : best), undefined);

function resolveOfficial(officialEvent: Event, weightBySeverity: Map<string, number>): Resolution {
  const severity = officialEvent.severity!;
  const total = [...weightBySeverity.values()].reduce((a, b) => a + b, 0);
  const agree = weightBySeverity.get(severity) ?? 0;
  return { severity, conflicted: false, bucket: 'official', confidence: total > 0 ? agree / total : 1, maxSeverity: severity };
}

function resolveCrowd(weighted: Weighted[], weightBySeverity: Map<string, number>): Resolution {
  const withWeight = [...weightBySeverity.entries()].filter(([, w]) => w > WEIGHT_FLOOR).map(([s]) => s);
  if (withWeight.length === 0) return { severity: 'S0', conflicted: false, bucket: 'unverified', confidence: 0, maxSeverity: 'S0' };

  // Rule A: the highest claimed tier is the ceiling.
  const maxSeverity = maxBy(withWeight, severityOrdinal)!;

  // Rule B: only a lower tier with >= 2 proximate observers can pull the display down.
  const lowerEligible = withWeight
    .filter((s) => severityOrdinal(s) < severityOrdinal(maxSeverity))
    .filter((s) => weighted.filter((w) => w.severity === s && w.proximate).length >= DEESCALATION_COUNT);

  const dangerWeight = (weightBySeverity.get('S2') ?? 0) + (weightBySeverity.get('S3') ?? 0);
  const safeWeight = (weightBySeverity.get('S0') ?? 0) + (weightBySeverity.get('S1') ?? 0);
  // Rule C: genuine disagreement is shown, not averaged; a met Rule B bar pre-empts it.
  const conflicted = lowerEligible.length === 0 && dangerWeight >= CONFLICT_WEIGHT_THRESHOLD && safeWeight >= CONFLICT_WEIGHT_THRESHOLD;

  const resolved = conflicted ? 'SX' : lowerEligible.length > 0 ? maxBy(lowerEligible, severityOrdinal)! : maxSeverity;

  if (conflicted) return { severity: resolved, conflicted: true, bucket: 'unverified', confidence: 0, maxSeverity };

  // "Unverified: confidence < 0.35, OR a single report" — a lone observer never reads as more.
  if (weighted.length <= 1) {
    const w = weightBySeverity.get(resolved);
    return { severity: resolved, conflicted: false, bucket: 'unverified', confidence: w === undefined ? 0 : wilsonLowerBound(w, 0), maxSeverity };
  }

  const agree = weightBySeverity.get(resolved) ?? 0;
  const disagree = [...weightBySeverity.entries()].filter(([s]) => s !== resolved).reduce((a, [, w]) => a + w, 0);
  const confidence = wilsonLowerBound(agree, disagree);
  const proximateAgreeing = weighted.filter((w) => w.severity === resolved && w.proximate).length;
  const bucket: Bucket = confidence >= 0.65 && proximateAgreeing >= 2 ? 'confirmed' : confidence >= 0.35 ? 'likely' : 'unverified';
  return { severity: resolved, conflicted: false, bucket, confidence, maxSeverity };
}

/**
 * Pure fold over one spot's events into its display state. `allEvents` is expected newest first, as
 * the phone's own query returns them; only an edge case (a spot with no flood_report left) reads
 * the order.
 */
export function summarize(featureRef: string, allEvents: Event[], now: number): FeatureSummary | null {
  return fold(featureRef, allEvents.filter((e) => e.featureRef === featureRef), now);
}

function fold(featureRef: string, featureEvents: Event[], now: number): FeatureSummary | null {
  if (featureEvents.length === 0) return null;
  const events = liveEvents(featureEvents);
  // Nothing with a reading is left: the spot leaves the map instead of showing an S0 nobody reported.
  if (!events.some((e) => e.severity != null)) return null;

  // One live position per author, so nobody stacks repeated confirms to fake consensus.
  const byAuthor = new Map<string, Event[]>();
  for (const e of events) byAuthor.set(e.authorId, [...(byAuthor.get(e.authorId) ?? []), e]);
  const latestPerAuthor = [...byAuthor.values()].map((es) => maxBy(es, (e) => e.timestampMs)!);

  // Anchor: the flood_reports' centroid, never a confirm's own position.
  const reports = events.filter((e) => e.type === 'flood_report');
  const avg = (xs: number[]) => xs.reduce((a, b) => a + b, 0) / xs.length;
  const anchorLat = reports.length ? avg(reports.map((e) => e.lat)) : events[0].lat;
  const anchorLon = reports.length ? avg(reports.map((e) => e.lon)) : events[0].lon;

  const weighted: Weighted[] = [];
  for (const event of latestPerAuthor) {
    if (event.severity == null) continue;
    const ageMs = Math.max(0, now - event.timestampMs);
    const proximity = proximityFactor(haversineMeters(anchorLat, anchorLon, event.lat, event.lon));
    weighted.push({
      event,
      severity: event.severity,
      weight: roleWeight(event.authorRole) * proximity * recencyFactor(ageMs, event.severity),
      proximate: proximity >= 0.7,
    });
  }
  // An official's own ruling is not also counted inside the crowd it overrides.
  const crowdWeighted = weighted.filter((w) => w.event.authorRole !== 'official');
  const weightBySeverity = new Map<string, number>();
  for (const w of crowdWeighted) weightBySeverity.set(w.severity, (weightBySeverity.get(w.severity) ?? 0) + w.weight);

  const crowd = resolveCrowd(crowdWeighted, weightBySeverity);

  // Rule D: an official event overrides the crowd, with the second-official gate as its one exception.
  const officialEvents = events.filter((e) => e.authorRole === 'official' && e.severity != null);
  const latestOfficial = maxBy(officialEvents, (e) => e.timestampMs);
  const officialSeverity = latestOfficial?.severity ?? null;

  const contradicting = officialSeverity
    ? crowdWeighted.filter((w) => severityOrdinal(w.severity) > severityOrdinal(officialSeverity) && w.weight > WEIGHT_FLOOR).length
    : 0;

  // The gate only holds back a de-escalation of a contradicted spot.
  const isDeEscalation = officialSeverity != null && severityOrdinal(officialSeverity) < severityOrdinal(crowd.maxSeverity);
  const gated = isDeEscalation && contradicting >= DEESCALATION_COUNT;
  const backingOfficials = new Set(officialEvents.filter((e) => e.severity === officialSeverity).map((e) => e.authorId)).size;
  const officialInForce = latestOfficial !== undefined && (!gated || backingOfficials >= REQUIRED_OFFICIALS_TO_DEESCALATE);

  const resolution = officialInForce ? resolveOfficial(latestOfficial!, weightBySeverity) : crowd;

  return {
    featureRef,
    lat: anchorLat,
    lon: anchorLon,
    severity: resolution.severity,
    confidence: resolution.confidence,
    bucket: resolution.bucket,
    isConflicted: resolution.conflicted,
    lastEventMs: Math.max(...events.map((e) => e.timestampMs)),
    isStale: now > Math.max(...events.map((e) => e.expiresAt)),
    confirmCount: events.filter((e) => e.type === 'confirm').length,
    disputeCount: events.filter((e) => e.type === 'dispute').length,
    officialSeverity,
    officialAuthorName: latestOfficial?.authorName ?? null,
    officialAtMs: latestOfficial?.timestampMs ?? null,
    pendingSecondOfficial: latestOfficial !== undefined && gated && !officialInForce,
    contradictingCount: contradicting,
    events: [...events, ...featureEvents.filter((e) => e.type === TYPE_FLOOD_WITHDRAW)].sort((a, b) => b.timestampMs - a.timestampMs),
  };
}

/** One summary per spot, grouping the log once instead of filtering it per spot. Events with no spot are ignored. */
export function summarizeAll(allEvents: Event[], now: number): FeatureSummary[] {
  const sorted = [...allEvents].sort((a, b) => b.timestampMs - a.timestampMs);
  const groups = new Map<string, Event[]>();
  for (const e of sorted) {
    if (!e.featureRef) continue;
    const g = groups.get(e.featureRef);
    if (g) g.push(e);
    else groups.set(e.featureRef, [e]);
  }
  return [...groups].map(([ref, evs]) => fold(ref, evs, now)).filter((s): s is FeatureSummary => s !== null);
}
